library/pk.c

```diff
@@ -877,6 +877,102 @@ mbedtls_pk_type_t mbedtls_pk_get_type(const mbedtls_pk_context *ctx)
     return ctx->pk_info->type;
 }
 
+#if defined(MBEDTLS_PSA_CRYPTO_CLIENT)
+int mbedtls_pk_copy_from_psa(mbedtls_svc_key_id_t key_id, mbedtls_pk_context *pk)
+{
+    psa_status_t status;
+    psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_type_t key_type;
+    psa_algorithm_t alg_type;
+    size_t key_bits;
+    /* Use a buffer size large enough to contain either a key pair or public key. */
+    unsigned char exp_key[PSA_EXPORT_KEY_PAIR_MAX_SIZE];
```

Comment:
Are we sure this is always large enough? For ECC keys, a key pair (unlike the name suggests) is exported as only the private key (one integer), so it's actually smaller than just a public key for Short Weierstrass curves (two integers + 1 byte). I seem to remember having this discussion before so we might have a macro for that already.

@gilles-peskine-arm This is arguably a bit of a gotcha in the PSA Crypto API, as it seems natural to assume a "pair" is larger than one of its components. Should we actually add a macro like `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE` (or a better name) to the official PSA API?

Reply:
> This is arguably a bit of a gotcha in the PSA Crypto API, as it seems natural to assume a "pair" is larger than one of its components.

Indeed. The key pair object is “larger” in terms of carrying strictly more information, but the representation doesn't have to be larger. Perhaps “key pair” objects should have been called “private key” objects, but that also as downsides, in particular that it's not obvious that every private key object implicitly includes a public key.

> `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE`

I suppose we could. It's easy enough to implement. I don't see a lot of value to it: it's uncommon that you'd know you have a key pair or a public key, but you don't know which. I guess that would only come up in key management APIs, which is sort of what we're doing here.

Reply:
I'm working on `mbedtls_pk_import_into_psa` and naturally I want to do exactly the same thing. What we're doing is a bit niche, but for that niche, `PSA_EXPORT_KEY_PAIR_OR_PUBLIC_MAX_SIZE` would indeed be convenient.